### ISSUE TITLE
[QUESTION] Best practices on imputation using Darts?

### ISSUE THREAD

**Issue by Tristanjmeyers** [unit8co/darts#2867]
Imputation
Hi!
I've created a simple xgboost model with Darts and I'm wanting to use this model to impute the same time series it was trained on. I know that there are some native ways in Darts to handle this with fill missing data and interpolation, but my series is a bit more complicated and actually requires another model to be trained to impute.

So, I'm looking for something like this:

```
import darts
from darts.models import XGBModel

## assume "targets" is a darts.timeseries.TimeSeries object and has some missing data

darts.utils.missing_values.missing_values_ratio(targets)
model = XGBModel(lags = 1, use_static_covariates=False)
target_subsets = darts.utils.missing_values.extract_subseries(targets, min_gap_size=1, mode='all')
model.fit(target_subsets)

targets_nogaps = model.impute(targets)
```

**Comment by eschibli** [comment]
darts does not have a forecaster-based imputer (though perhaps it should). If you need to use a forecaster I'd approach the task like this: 

1. Save a mask indicating the locations of the nan values, then replace the nan values with an indicator value (necessary as all models implimented will throw an exception if there are nan targets)
2. Train your imputing model (it will need to be robust to NaNs in features) with zero sample weights for the indicator-replaced NaNs.  
3. Create a single historical forecast **without retraining**
4. Extract the predictions for the values where you are missing the original data. 

Something like eg, 

```
def generate_noisy_sine_wave(n_samples=1000, noise_level=0.1, offset=0.1):
    t = np.linspace(0, 100, n_samples)
    y = np.sin(t) + noise_level * np.random.randn(n_samples) + offset
    return pd.Series(y, index=pd.date_range(start='2020-01-01', periods=n_samples, freq='D'))

noisy_series = generate_noisy_sine_wave()

# Randomly assign NaN to 10% of the data
nan_indices = np.random.choice(noisy_series.index, size=int(0.1 * len(noisy_series)), replace=False)
noisy_series.loc[nan_indices] = np.nan

# Create a mask to indicate the NaN values
nan_mask = noisy_series.isna()

# Save the original series (with NaNs) as a TimeSeries
og_ts = TimeSeries.from_series(noisy_series)

# Fill NaNs with an indicator value for model fitting (they will be deweighted)
noisy_series_filled = noisy_series.fillna(-1E6)
ts = TimeSeries.from_series(noisy_series_filled)

# Create a sample weight series to prevent training on the indicators
sample_weights = (~nan_mask).astype(float)
weights_ts = TimeSeries.from_series(sample_weights)

# Define and fit the XGB model using the filled series and sample weights
model = XGBModel(lags=24, output_chunk_length=1)
model.fit(ts, sample_weight=weights_ts)

# Generate historical forecasts without retraining
hf = model.historical_forecasts(series=ts, retrain=False)

# Use the forecast only where the original data had NaNs; otherwise, keep the original values
series_imputations = hf.to_series().where(nan_mask)
series_og = og_ts.to_series()
series_imputed = series_imputations.fillna(series_og)

# Convert the imputed series back to a TimeSeries

# Observed + imputed points
imp_ts = TimeSeries.from_series(series_imputed)

# Just the imputed points
imp_points_ts = TimeSeries.from_series(series_imputations)

imp_ts[-100:].plot(linewidth=1, label='Imputed Series', color='green')
og_ts[-100:].plot(linewidth=3, marker='o', markersize=3, label='Original Series', color='blue')
imp_points_ts[-100:].plot(marker='o', markersize=3, linestyle='', label='Imputed Points', color='orange')
plt.show()
```

<img width="554" height="431" alt="Image" src="https://github.com/user-attachments/assets/7c14ffdd-f08b-47c5-b261-0d449909f8e3" />

Caveats: If your data is monovariate (and honestly even if it's not) you will probably have better results with simple interpolation methods (but you said that won't work for you). If this is because you need to consider covariates, but are able to interpolate rather than just forecast, if performance is critical, you would probably see the best results with some sort of autoencoder. 

Also, simply using `output_chunk_length=1` *may* give poor performance if there are runs of adjacent nans (as the highly weighted recent observations will be missing), and you *might* get better performance with a more artisanal approach that fills in the best available prediction. I would highly recommend against setting `lags=1` however as that will result in highly degraded performance in such cases (simply predicting the training set mean in the case of two adjacent nans if there is no covariate information)

**Comment by Tristanjmeyers** [comment]
Thanks so much, this is perfect! I had no idea about the 0 sample weights feature. My approach actually used a base linear interpolation between missing points, then built a model to learn the whole series. I used the same masking method as you. But your method is much better as it doesn't just learn the linear interpolation., e.g.:


```   
# initial definition.
model = XGBModel(lags = 1, use_static_covariates=False)

# temporarily fill data will linear in order to get an xgboost model trained. 
tseries_filled = darts.utils.missing_values.fill_missing_values(tseries, method="linear")
model.fit(tseries_filled)
    
# Create a full prediction of the dataset
imputations = model.predict(len(tseries), series=tseries)

filled_vals = tseries.values(copy=True)

nan_mask = np.isnan(filled_vals)

# fill where nans are
filled_vals[nan_mask] = imputations.values()[nan_mask]

# Generate the timeseries dataset. 
target_imputed = TimeSeries.from_times_and_values(tseries.time_index, filled_vals, columns=tseries.components, static_covariates= tseries.static_covariates)
 
assert darts.utils.missing_values.missing_values_ratio(target_imputed) == 0, "There are missing values after imputation."
```


Also, great point about lags=1. For context, I'm trying to make the imputation robust and have as minimal dependencies as possible, hence only one lag point is required. But you're right, this has some significant trade-offs.

**Comment by dennisbader** [comment]
Thanks for another great answer and proposed solution @eschibli! Using `historical_forecasts()` with a pre-trained model is definitely the preferred option over an auto-regressive forecast with `predict()`.

I'll close this issue :)
